core/web/resolver/job.go

```diff
@@ -83,31 +84,55 @@ func (r *JobResolver) Spec() *SpecResolver {
 	return NewSpec(r.j)
 }
 
-func (r *JobResolver) Runs(ctx context.Context) ([]*JobRunResolver, error) {
-	runs, err := loader.GetJobRunsByPipelineSpecID(ctx, strconv.Itoa(int(r.j.PipelineSpecID)))
+// Runs fetches the runs for a Job.
+func (r *JobResolver) Runs(ctx context.Context, args struct {
+	Offset *int32
+	Limit  *int32
+}) (*JobRunsPayloadResolver, error) {
+	offset := pageOffset(args.Offset)
+	limit := pageLimit(args.Limit)
+
+	//
+	if limit > 100 {
+		limit = 100
+	}
+
+	ids, err := r.app.JobORM().FindPipelineRunIDsByJobID(r.j.ID, offset, limit)
```

Comment:
This change introduces a new problem; now our ID fetch, load and count are all done in separate transactions. So if a job run is deleted or added the numbers will be wrong - it's not atomic.

Wrapping this load here in a transaction feels messy.

I would rather see one ORM method that encapsulates everything that you're trying to do instead of splitting it into three.

Reply:
In GraphQL you declare the fields you want and it will only execute the resolvers of those fields. For instance if you call

```
job(id: 1) {
  id
}
```

This will only make a DB call to fetch the job without having to preload any relationships (although in our case we still do because that's how the ORM methods are currently set up).

If you wanted to get the runs it you would do

```
job(id: 1) {
  id
  runs {
    results {
      id
    }
    metadata {
      count
    }
  }
}
```

This would execute the run resolvers which would make the new DB calls.

This is the idea of dataloaders in GQL which allows you to only query for what is requested rather than preloading everything up front. Understandably the atomicity of the query is impacted, but I think this is a decent tradeoff for the API which is used for the UI.

Reply:
Another thought is about is how the UI is currently dealing with displaying runs for a job.

Currently the REST API does not support embedded runs inside is response body. In order to display the runs for a job, we make a separate REST API call to fetch the job runs for a job (`jobs/:id/runs`), which would also be run in a separate transaction.